modules/chemical_reactions/src/utils/ReactionNetworkUtils.C

```diff
@@ -0,0 +1,308 @@
+      return term;
+    }
+    else
+      return std::any_cast<Term>(vs[0]);
+  };
+
+  parser["TermList"] = [](const SemanticValues & vs)
+  {
+    TermList terms;
+    for (const auto & v : vs)
+    {
+      // One term
+      if (v.type() == typeid(Term))
+        terms.push_back(std::any_cast<Term>(v));
+      // Multiple terms
+      else if (v.type() == typeid(std::vector<Term>))
+      {
+        const auto & more = std::any_cast<std::vector<Term>>(v);
+        terms.insert(terms.end(), more.begin(), more.end());
+      }
+    }
+    return terms;
+  };
+
+  // Leaf types can just use 'vs.token()'
+  parser["Key"] = [](const SemanticValues & vs) { return vs.token(); };
+
+  parser["Value"] = [](const SemanticValues & vs) { return vs.token(); };
+
+  parser["Pair"] = [](const SemanticValues & vs)
+  {
+    mooseAssert(vs.size() == 2, "Metadata entry should have exactly one key and one value");
+    return std::make_pair(std::string(std::any_cast<std::string_view>(vs[0])),
+                          std::string(std::any_cast<std::string_view>(vs[1])));
+  };
+
+  parser["Metadata"] = [](const SemanticValues & vs)
+  {
+    Metadata meta;
+    for (const auto & v : vs)
+    {
+      if (v.type() == typeid(std::pair<std::string, std::string>))
+      {
+        const auto & [k, val] = std::any_cast<std::pair<std::string, std::string>>(v);
+        meta[k] = val;
+      }
+      else if (v.type() == typeid(std::vector<std::pair<std::string, std::string>>))
+      {
+        const auto & pairs = std::any_cast<std::vector<std::pair<std::string, std::string>>>(v);
+        for (const auto & [k, val] : pairs)
+          meta[k] = val;
+      }
+    }
+    return meta;
+  };
+
+  parser["Reaction"] = [](const SemanticValues & vs)
+  {
+    Reaction r;
+    if (vs.size() != 2 && vs.size() != 3)
+      mooseError("Reaction failed to read as 'reactors -> products [metadata]");
+    r.reactants = std::any_cast<TermList>(vs[0]);
+    r.products = std::any_cast<TermList>(vs[1]);
+    if (vs.size() == 3)
+      r.metadata = std::any_cast<Metadata>(vs[2]);
+    return r;
+  };
+
+  parser.enable_packrat_parsing();
+
+  std::vector<Reaction> reactions;
+  if (!parser.parse(reaction_network_string, reactions))
+    mooseError("Failed to parse reaction.");
+
+  // Output the reaction network
+  if (output_to_cout)
+  {
+    for (size_t i = 0; i < reactions.size(); ++i)
+    {
+      const auto & r = reactions[i];
+      Moose::out << "Reaction " << i + 1 << ":\n";
+      Moose::out << Moose::stringify(r) << std::endl;
+    }
+  }
+
+  return reactions;
+}
+
+std::vector<VariableName>
+Reaction::getSpecies() const
+{
+  std::vector<VariableName> species;
+  for (const auto & term : reactants)
+    species.push_back(term.species + term.state.value_or("") + term.charge.value_or(""));
+  for (const auto & term : products)
+    species.push_back(term.species + term.state.value_or("") + term.charge.value_or(""));
+  return species;
+}
```

Comment:
Clarifying question:
As rewritten, this allows for duplicate species names in a single reaction, correct? For example, someone inputs:
`e + Ar -> e + e + Ar+` for argon ionization, this would have three entries for electrons, `e` (one entry for reactions and two entries for products).

If so, that is good thing. I just wanted to point out that we might want to include a "consolidated function" here, or maybe just have the inheriting objects handle that?

Reply:
Do you think it is worth having a "consolidated function" function here, to keep track of the total Stoichiometric coefficient, so should that be the job of the inheriting objects?